File: QueueWithStack.py
# Time Complexity : 28
#  Space Complexity : 14
#  Did this code successfully run on Leetcode : Yes
#  Any problem you faced while coding this : For one case I had wrong output as the else condition was not handled correctly


#  Your code here along with comments explaining your approach
class MyQueue:

    #Define two stacks to manipulate them as queue and make the stack reverse in order
    def __init__(self):
        self.stackA = []
        self.stackB = []

    #append elements in first stack
    def push(self, x: int) -> None:
        self.stackA.append(x)
        
    #Reversing the stack by popping the first and appending it to second, return the element at top of second stack
    def pop(self) -> int:
        if not self.stackB:
            while self.stackA:
                self.stackB.append(self.stackA.pop())
        return self.stackB.pop()
    
    #Repeats the same process of popping and pushin the element to second stack, returns the top element of second stack
    def peek(self) -> int:
        if not self.stackB:
            while self.stackA:
                self.stackB.append(self.stackA.pop())
        return self.stackB[-1]
    
    #if both stacks are empty return true
    def empty(self) -> bool:
        return not self.stackA and not self.stackB


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()

File: QueueWithStack.java
// Time Complexity :O(1) Amortized
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes

import java.util.Stack;

public class QueueWithStack {
  Stack<Integer> s1;
  Stack<Integer> s2;

  public QueueWithStack() {
    s1 = new Stack<Integer>();
    s2 = new Stack<Integer>();
  }

  public void push(int x) {
    s1.push(x);
  }

  public int pop() {
    if (!s2.isEmpty()) {
      return s2.pop();
    } else {
      while (!s1.isEmpty()) {
        int temp = s1.peek();
        s1.pop();
        s2.push(temp);
      }
      return s2.pop();
    }
  }

  public int peek() {
    if (s2.size() != 0) {
      return s2.peek();
    } else {
      while (!s1.isEmpty()) {
        int temp = s1.pop();
        s2.push(temp);
      }
      return s2.peek();
    }
  }

  public boolean empty() {
    if (!s1.isEmpty() || !s2.isEmpty()) return false;
    return true;
  }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * boolean param_4 = obj.empty();
 */
